designs/switch-prometheus-to-opentelemetry.md

```diff
@@ -0,0 +1,172 @@
+Switch Prometheus to OpenTelemetry
+===================
+
+## Motivation
+
+Today, controller-runtime already provides metrics accessible through the
+Prometheus client, and exposed on the `/metrics` endpoint. However, this forces
+components to use Prometheus, when they may want to monitor their app another
+way, not necessarily with a pull system.
+
+## Goals
+
+* Provide a generic way to send metrics, so they can be exported virtually anywwhere.
```

Comment:
I do not find this very convincing in its current form. Prometheus metrics format is not only supported by Prometheus itself, but also for example by [gcp](https://cloud.google.com/stackdriver/docs/solutions/gke/prometheus) and [aws](https://aws.amazon.com/about-aws/whats-new/2020/05/amazon-cloudwatch-now-monitors-prometheus-metrics-now-in-beta/).

Another option to solve your problem would be a sidecar that scrapes the prometheus metrics and puts them into whatever system you have. Have you considered that? If you think it is not suited, please add that into an `Alternatives considered` section and explain why you think it is problematic.

Could you describe some more concrete examples of why the options above are not enough? Is there prior art of projects in the kube ecosystem going this way?

Reply:
It's not that those options aren't enough. It's that not using Prometheus allows more flexibility on the side of applications.
OpenTelemetry definitely allows using the Prometheus metrics format. But also any other format if the user of the library wishes to.

As for a sidecar, yes, it's always a possibility. But also an added reliability risk, as well as complexity.

Reply:
> I do not find this very convincing in its current form. Prometheus metrics format is not only supported by Prometheus itself, but also for example by [gcp](https://cloud.google.com/stackdriver/docs/solutions/gke/prometheus) and [aws](https://aws.amazon.com/about-aws/whats-new/2020/05/amazon-cloudwatch-now-monitors-prometheus-metrics-now-in-beta/).

Those cloud providers are all-in on OpenTelemetry, too.

- https://cloud.google.com/trace/docs/setup "Cloud Trace recommends using OpenTelemetry."
- https://github.com/GoogleCloudPlatform/opentelemetry-operations-go "OpenTelemetry exporters to Google Cloud Trace and Google Cloud Monitoring."
- https://aws.amazon.com/otel "instrument your applications just once to send correlated metrics and traces to multiple AWS and Partner monitoring solutions"
- https://github.com/aws-observability/aws-otel-collector "send telemetry data to AWS CloudWatch Metrics, Traces and Logs"

> Is there prior art of projects in the kube ecosystem going this way?

sig-instrumentation intends to move forward with it once they can resolve gRPC dependencies with their exporter of choice: [https://feature.k8s.io/647](https://git.k8s.io/enhancements/keps/sig-instrumentation/647-apiserver-tracing)

> OpenTelemetry definitely allows using the Prometheus metrics format. But also any other format if the user of the library wishes to.

By instrumenting with OpenTelemetry, this module (controller-runtime) can import a lightweight and stable API and leave the choice of exporter/format to applications and deployers.

Here's a list of existing exporters: https://opentelemetry.io/registry/?language=go&component=exporter

<details><summary>Here are its instrumentation dependencies: <strong>tl;dr</strong> stdlib, go-cmp, and testify.</summary>

```
● /go/src/go.opentelemetry.io/otel (v0.18.0) go mod graph
go.opentelemetry.io/otel github.com/google/go-cmp@v0.5.4
go.opentelemetry.io/otel github.com/stretchr/testify@v1.7.0
go.opentelemetry.io/otel go.opentelemetry.io/otel/metric@v0.18.0
go.opentelemetry.io/otel go.opentelemetry.io/otel/oteltest@v0.18.0
go.opentelemetry.io/otel go.opentelemetry.io/otel/trace@v0.18.0
github.com/google/go-cmp@v0.5.4 golang.org/x/xerrors@v0.0.0-20191204190536-9bdfabe68543
github.com/stretchr/testify@v1.7.0 github.com/davecgh/go-spew@v1.1.0
github.com/stretchr/testify@v1.7.0 github.com/pmezard/go-difflib@v1.0.0
github.com/stretchr/testify@v1.7.0 github.com/stretchr/objx@v0.1.0
github.com/stretchr/testify@v1.7.0 gopkg.in/yaml.v3@v3.0.0-20200313102051-9f266ea9e77c
go.opentelemetry.io/otel/metric@v0.18.0 github.com/google/go-cmp@v0.5.4
go.opentelemetry.io/otel/metric@v0.18.0 github.com/stretchr/testify@v1.7.0
go.opentelemetry.io/otel/metric@v0.18.0 go.opentelemetry.io/otel@v0.18.0
go.opentelemetry.io/otel/metric@v0.18.0 go.opentelemetry.io/otel/oteltest@v0.18.0
go.opentelemetry.io/otel/oteltest@v0.18.0 go.opentelemetry.io/otel@v0.18.0
go.opentelemetry.io/otel/oteltest@v0.18.0 go.opentelemetry.io/otel/metric@v0.18.0
go.opentelemetry.io/otel/oteltest@v0.18.0 go.opentelemetry.io/otel/trace@v0.18.0
go.opentelemetry.io/otel/trace@v0.18.0 github.com/stretchr/testify@v1.7.0
go.opentelemetry.io/otel/trace@v0.18.0 go.opentelemetry.io/otel@v0.18.0
gopkg.in/yaml.v3@v3.0.0-20200313102051-9f266ea9e77c gopkg.in/check.v1@v0.0.0-20161208181325-20d25e280405
go.opentelemetry.io/otel@v0.18.0 github.com/google/go-cmp@v0.5.4
go.opentelemetry.io/otel@v0.18.0 github.com/stretchr/testify@v1.7.0
go.opentelemetry.io/otel@v0.18.0 go.opentelemetry.io/otel/metric@v0.18.0
go.opentelemetry.io/otel@v0.18.0 go.opentelemetry.io/otel/oteltest@v0.18.0
go.opentelemetry.io/otel@v0.18.0 go.opentelemetry.io/otel/trace@v0.18.0
```

If we were to wrap our `http.RoundTripper` in the Kubernetes client to automatically trace/propagate our calls to the API server, it would add one more module: `github.com/felixge/httpsnoop`

```
● /go/src/go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp (v0.18.0) go mod graph | wc -l
27
```

Prometheus...

```
● /go/pkg/mod/github.com/prometheus/client_golang@v1.9.0 -- go mod graph | wc -l
828
```

</details>
<br/>

Reply:
> Those cloud providers are all-in on OpenTelemetry, too.

That isn't really a pro OpenTelemetry argument IMHO, that is a "Cloud providers are fine with either"

> sig-instrumentation intends to move forward with it once they can resolve gRPC dependencies with their exporter of choice: https://feature.k8s.io/647

That kep is about tracing, not metrics and hasn't made it anywhere yet.

> By instrumenting with OpenTelemetry, this module (controller-runtime) can import a lightweight and stable API and leave the choice of exporter/format to applications and deployers.

Yeah but I think it is important to provide a sane default and not put that burden onto the user. Prometheus is a sane default. Even if we were to switch to Open Telemetry we should still have Prometheus as default like we do today.

I guess my feedback on this is: This needs a compelling user story. Something you can not at all or only with significant issues do today. Changing the libs in here will force most users of c-r to change their code and I do not want to do that without also being certain that this is a significant value add. "Cloud providers also accept open telemetry" and "if you want you can now use something other than Prometheus" without a single concrete example of that just isn't very convincing and makes this appear to me like a change for the sake of change, not for improving things.

Reply:
> > By instrumenting with OpenTelemetry, this module (controller-runtime) can import a lightweight and stable API and leave the choice of exporter/format to applications and deployers.
> 
> Yeah but I think it is important to provide a sane default and not put that burden onto the user. Prometheus is a sane default. Even if we were to switch to Open Telemetry we should still have Prometheus as default like we do today.

👍🏻 

> I guess my feedback on this is: This needs a compelling user story. Something you can not at all or only with significant issues do today. Changing the libs in here will force most users of c-r to change their code and I do not want to do that without also being certain that this is a significant value add.

My impression was there there was only one mention of Prometheus at [`manager.Options.MetricsBindAddress`](https://github.com/kubernetes-sigs/controller-runtime/blob/774f9d490254c861e9f5ca009b3c4b8b5d55fa83/pkg/manager/manager.go#L196-L199), but now I see that there's also a `Manager.AddMetricsExtraHandler()` method and an exported `prometheus.Registerer` in `./pkg/metrics`. I wouldn't expect these to survive a transition away from Prometheus, so they do seem like blockers.

@dmathieu I think these would need to be addressed in the proposal.

> "Cloud providers also accept open telemetry" and "if you want you can now use something other than Prometheus" without a single concrete example of that just isn't very convincing and makes this appear to me like a change for the sake of change, not for improving things.

My sense is that the OpenTelemetry API is better suited for packages like this because of its attention to decoupling instrumentation from pushing/pulling metrics. The microscopic number of dependencies is one symptom of that. But it's just a hunch at this point; I haven't done more than toy with it.

So, I agree.